v2/protocol/pubsub/message.go

```diff
@@ -0,0 +1,123 @@
+package pubsub
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"strings"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/binding/format"
+	"github.com/cloudevents/sdk-go/v2/binding/spec"
+)
+
+const (
+	prefix      = "ce-"
+	contentType = "Content-Type"
+)
+
+var specs = spec.WithPrefix(prefix)
+
+// Message represents a Pub/Sub message.
+// This message *can* be read several times safely
+type Message struct {
+	Msg pubsub.Message
+
+	format   format.Format
+	version  spec.Version
+	encoding binding.Encoding
+}
+
+// NewMessage returns a binding.Message with data and attributes.
+// This message *can* be read several times safely
+func NewMessage(data []byte, attributes map[string]string) *Message {
+	encoding := binding.EncodingBinary
+	var f format.Format = nil
+	var version spec.Version = nil
+	if attributes != nil {
+		// Use Content-type attr to determine if message is structured and
+		// set format.
+		if s := attributes[contentType]; format.IsFormat(s) {
+			f = format.Lookup(s)
+			encoding = binding.EncodingStructured
+		}
+		// Binary v0.3:
+		if s := attributes[prefix+"specversion"]; s != "" {
+			version = specs.Version(s)
+		}
+	}
+
+	// Check as Structured encoding.
+	raw := make(map[string]json.RawMessage)
+	if err := json.Unmarshal(data, &raw); err != nil {
+		version = specs.Version("")
+	}
+
+	// Structured v0.3
+	if v, ok := raw["specversion"]; ok {
+		var ver string
+		if err := json.Unmarshal(v, &ver); err != nil {
+			version = specs.Version("")
+		}
+		version = specs.Version(ver)
```

Comment:
Why do you unmarshal the json to read the version? you don't need that

Reply:
this was code that I ported from the previous implementation of message.go. Removed it.